corehq/apps/locations/util.py

```diff
@@ -179,55 +199,68 @@ def get_consumption(self, loc):
-        _extend_headers('uncategorized_data', uncategorized_keys)
-        if self.include_consumption_flag and loc_type.name not in self.administrative_types:
-            _extend_headers('consumption', self.product_codes)
-
-        sheet_title = loc_type.code
-
-        return (sheet_title, {
-            'headers': tab_headers,
-            'rows': tab_rows,
-        })
+    def get_headers(self):
+        headers = OrderedDict([
+            ('types', [list(LOCATION_TYPE_SHEET_HEADERS.values())])
+        ])
+        for loc_type in self.location_types:
+            additional_headers = []
+            additional_headers.extend(self._prefix_headers('data', (f.slug for f in self.data_model.fields)))
+            if self.include_consumption_flag and loc_type.name not in self.administrative_types:
+                additional_headers.extend(self._prefix_headers('consumption', self.product_codes))
+            additional_headers.append(LOCATION_SHEET_HEADERS_OPTIONAL['uncategorized_data'])
+            additional_headers.append(LOCATION_SHEET_HEADERS_OPTIONAL['delete_uncategorized_data'])
+
+            headers[loc_type.code] = [LOCATION_SHEET_HEADERS_BASE.values() + additional_headers]
+
+        return list(headers.items())
+
+    def write_data(self, writer):
+        self._write_type_sheet(writer)
+        for location_type in self.location_types:
+            self._write_locations(writer, location_type)
+
+    def _write_locations(self, writer, location_type):
+        include_consumption = self.include_consumption_flag and location_type.name not in self.administrative_types
+        query = SQLLocation.active_objects.filter(
+            domain=self.domain,
+            location_type__name=location_type.name
+        )
 
-    def type_sheet(self, location_types):
-        headers = LOCATION_TYPE_SHEET_HEADERS
+        def _row_generator(include_consumption=include_consumption):
+            for loc in query:
+                model_data, uncategorized_data = self.data_model.get_model_and_uncategorized(loc.metadata)
+                row = [
+                    loc.location_id,
+                    loc.site_code,
+                    loc.name,
+                    loc.parent.site_code if loc.parent else '',
+                    loc.latitude or '',
+                    loc.longitude or '',
+                    loc.external_id,
+                    '',  # do delete
+                ]
+                for field in self.data_model.fields:
+                    row.append(model_data.get(field.slug, ''))
+
+                if include_consumption:
+                    consumption_data = self.get_consumption(loc)
+                    row.extend([consumption_data[code] for code in self.product_codes])
+
+                row.append(', '.join('{}: {}'.format(*d) for d in uncategorized_data.items()))
+
+                yield row
+                self._increment_progress()
+
+        writer.write([(location_type.code, _row_generator())])
```

Comment:
Nice, so then `writer` will stream the generator without holding the whole thing in memory?

Reply:
Correct